Port, Update and Merge Triflight Improved Tricopter Control

SJ-Innovation opened the pull request:
Following the stagnation of the Triflight Fork of Cleanflight it made sense to take the changes that were made to the Tricopter mixer and update them for merge into Betaflight. 

This replaces the stock Tricopter mixer, and defaults to a feedback-less servo (virtual feedback simulation). When a hardware ADC is used for servo feedback, this improves the responce of the servo. 

Unit Tests are also included to test as much of the lower level calculation functions as possible. 

More to come: Just wanted to create a PR to run through CI pipelines for tests.

SJ-Innovation commented:
**Progress:**
Ported and Updated. Compiles and runs on MatekF405 boards. It runs! WOO! The Unit Tests also run and pass, proving that the underlying coefficient calculators are working as expected. Going to run some flight tests soon, once I have access to my `copters again in a few days. 

**Next Steps:**
Planning on adding a FEATURE_ to enable or disable the Triflight improved mixer (they take a reasonable amount of memory and are not useful for other machines other than tricopters, so would be handy to be disabled on builds that do not need them)

Planning to do more cleanups 

**Issue:**
Having some issues with some F7 (I think F722) targets due to size of ITCM RAM (16K), hense the current pipeline failure.
AlienWhoopF7 is my current bugbear.  

I know that the macros in `common_pre.h` such as `FAST_CODE` etc can be used to control  the location of variables and functions in memory, and that optimisations can be customised in the makefile (speed, size etc). But would also appreicate a little bit  of advice on the particular optimisations that should be applied to the mixer. 
At present there are 2 files for the tricopter mixer, one is speed optimised and contains the loop mixer code. The other is size optimised and contains the initialisation and less speed dependant functions. Does this seem sensible? 

Is there any way you know of that i can force the init code out of ITCM and into some other (slower but larger) flash? 

@mikeller Could you shed a little light? ( / Lend a suggestion or two as to how to proceed)

TheIsotopes commented:
@SJ-Innovation i would like to test this.
please can you rebase this to the current master.

SJ-Innovation commented:
I currently do not have an internet connection sufficient of anything more than emails. I am moving house at the moment and am rather busy.

I'll do this (and more, like actually finish the port...) Once I'm fully moved in and have internet.

On 13 Aug 2020 14:56, TheIsotopes <notifications@github.com> wrote:

@SJ-Innovation<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FSJ-Innovation&data=02%7C01%7Csl01172%40surrey.ac.uk%7C6e77425643ed417647c208d83f909492%7C6b902693107440aa9e21d89446a2ebb5%7C0%7C0%7C637329237761934037&sdata=gtbzrbK204Xy3GuB6wB8NA1oPq70ZeaqmY5iYBldIfs%3D&reserved=0> i would like to test this.
please can you rebase this to the current master.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbetaflight%2Fbetaflight%2Fpull%2F10051%23issuecomment-673493733&data=02%7C01%7Csl01172%40surrey.ac.uk%7C6e77425643ed417647c208d83f909492%7C6b902693107440aa9e21d89446a2ebb5%7C0%7C0%7C637329237761944034&sdata=GfNeFZlKRJrddz5r1fSMfGcRRqMtNIFFJZqC3OQKeug%3D&reserved=0>, or unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACJALZT6PQH6EVVSYIWKJ2DSAPWGDANCNFSM4PISEKLQ&data=02%7C01%7Csl01172%40surrey.ac.uk%7C6e77425643ed417647c208d83f909492%7C6b902693107440aa9e21d89446a2ebb5%7C0%7C0%7C637329237761944034&sdata=uH2HHuzUd%2BChV7J%2BuyE8nHpdsgzyXuLdIjj4bTGmIU4%3D&reserved=0>.

TheIsotopes commented:
@SJ-Innovation imo i found a bug
when `use_triflight` is activated, I cannot reverse the servo travel.
back without `use_triflight` everything is ok with the servo.

SJ-Innovation commented:
Rebased and updated. Currently depending on #10116. Currently F7 targets have ITCM disabled. This is definitely *not-ideal*...

@TheIsotopes Thank you for that. As i say this is currently untested by me (hardware going through some upgrades) following the house move. Once i get my hardware up and running fully again itll progress much more quickly. Im just a little busy ;)

TheIsotopes commented:
@SJ-Innovation could you please rearrange all of this.
I think things from #10116 (or the split) have been added here..

SJ-Innovation commented:
I'll rearrange it at the opportune time. I am currently in active development and negotiation with the senior merge-privileged members to get the parts of the split added as they are all prerequisites of this.
As they are merged I'll rebase and update periodically, safe to say I'm on top of it ;) But as with holding down a job, degree and personal life I don't spend all day doing this 😂 It'll get done as it gets done.

On 26 Aug 2020 08:49, TheIsotopes <notifications@github.com> wrote:

@SJ-Innovation<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FSJ-Innovation&data=02%7C01%7Csl01172%40surrey.ac.uk%7C9255c9d165544e66bc7708d849949c47%7C6b902693107440aa9e21d89446a2ebb5%7C0%7C0%7C637340249909117806&sdata=hWfPTVobGhIs21Wv%2FQz3lOhUUpiFrqFr%2FtGrNOb1FDA%3D&reserved=0> could you please rearrange all of this.
I think things from #10116<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbetaflight%2Fbetaflight%2Fpull%2F10116&data=02%7C01%7Csl01172%40surrey.ac.uk%7C9255c9d165544e66bc7708d849949c47%7C6b902693107440aa9e21d89446a2ebb5%7C0%7C0%7C637340249909127804&sdata=von9OZWIAtbnxSmSquPBwdwS2AIUFXxkQPHQygZFNFk%3D&reserved=0> (or the split) have been added here..

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbetaflight%2Fbetaflight%2Fpull%2F10051%23issuecomment-680719269&data=02%7C01%7Csl01172%40surrey.ac.uk%7C9255c9d165544e66bc7708d849949c47%7C6b902693107440aa9e21d89446a2ebb5%7C0%7C0%7C637340249909127804&sdata=RE%2FRUiYSci4dlFgelDikAw%2FEI5wggKh7VXt3vDXvimU%3D&reserved=0>, or unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACJALZWBAR4I2Z6CMAZ65RTSCS5BZANCNFSM4PISEKLQ&data=02%7C01%7Csl01172%40surrey.ac.uk%7C9255c9d165544e66bc7708d849949c47%7C6b902693107440aa9e21d89446a2ebb5%7C0%7C0%7C637340249909137795&sdata=zvsw8SA2%2BN9lNAuHd%2FEBzZzAEtkiESzv02dUI7cXr7E%3D&reserved=0>.

TheIsotopes commented:
@SJ-Innovation is there any news or progress on this PR?

SJ-Innovation commented:
@TheIsotopes Im working on it between my degree and other commitments. I'm also having to work around some flash size issues as ive been refused a merge on a pretty critical piece of memory prioritisation code that makes this PR a lot easier. Its happening though.

mikeller commented:
@SJ-Innovation: Realistically, seeing the diminutive number of users that triflight has, I think we'll only ever want to include this code for MCUs with 1 MB flash (F405, F745), so flash size should not be an issue here.

SJ-Innovation commented:
@mikeller Seems like a reasonable conclusion. I guess if anyone wants Triflight on an otherwise unsupported MCU they're probably more than capable of editing the config to include it.

TheIsotopes commented:
@SJ-Innovation @mikeller i could also make friendly with this solution.

sightseekerjb commented on src/main/blackbox/blackbox.c:
```diff
@@ -1881,7 +1881,7 @@ uint8_t blackboxCalculateSampleRate(uint16_t pRatio) {
-void blackboxInit(void)
+SLOW_CODE void blackboxInit(void)
 {
     blackboxResetIterationTimers();
```
hey any clue on cc3d anything new 2020

juvinski commented:
Hi,

I'm trying to flash your branch on a FuryF4OSD and after compiling, flashing the CLI is not working - I'm having strange characters like:
Õ$M>	n	ÿ[Ë
And I can't discovery what is happening.
Any clue ?

SJ-Innovation commented:
@juvinski Yeah it was an issue i  was occasionally having on my F4 boards. I reckon its down how things are allocated in memory (the main reason why this branch staled is that a key memory saving feature i needed got firmly rejected...).

Ive been meaning to have a new-found hash of this to sort a bit of other stuff out. In the meantime i recommend you check out @jihlein s fork. Hes been doing some internal triflight stuff (cant warranty nor vouch at this time though, im a little out of the loop)
